man/man8/zfsprops.8

```diff
@@ -962,6 +962,103 @@ Unless necessary, deduplication should NOT be enabled on a system. See the
 section of
 .Xr zfsconcepts 8 .
 .It Xo
+.Sy directio Ns = Ns Sy off Ns | Ns Sy on Ns | Ns Sy strict Ns | Ns
+.Sy legacy
+.Xc
+Sets the policy for Direct IO when an IO request uses the
+.Sx O_DIRECT
+flag. The default value is
+.Sy strict.
+.Pp
+All Direct IO requests must be page-aligned.
+.Pp
+If this property is set to
+.Sy off
+all Direct IO requests will always fail.
+.Pp
+If this property is set to
+.Sy on
+all IO requests requesting Direct IO will attempt to bypass the ARC
+based on the
+.Sx alignment
+of the IO request descripted below by
+.Sy directio_write_align
+and
+.Sy directio_read_align
+properties. In the event the IO request is not properly aligned based on
+the alignment property, the request will just be redirected to use the ARC.
+.Pp
+If this parameter is set to
+.Sy strict
+all IO requests requesting Direct IO will not fail as long as they follow
+the alignment property. However, if the alignment policy is not followed
+the IO request will fail.
+.Pp
+If this parameter is set to
+.Sy legacy
+all IO requests requesting Direct IO will always use the ARC. This is
+based on the original mode for the
+.Sx O_DIRECT
+flag in ZFS. This allows ZFS to accept and silently ignore
+.Sx O_DIRECT
+flag.
+.Pp
+.Sx NOTE:
+By default all
+.Sx O_DIRECT
+IO requests that are page-aligned are valid for Direct IO. However,
+this is not highly performant for
+.Sx writes .
+Any page-aligned Direct IO
+.Sx write
+page-aligned request that is not also recordsize-aligned will requires
+read, write, and modify cycle. This can cause quite a large
+performance penalty to occur. It is suggested to set this property to
```

Comment:
I'm wondering if all of the different settings are actually useful in different circumstances, or are we just giving users the opportunity to misconfigure the system.

Given the performance impact of non-recordsize-aligned DIRECTIO writes, could we change the behavior so that these non-recordsize-aligned writes will always use the "normal" (cached, non-directio) behavior?

Reply:
Yes, I was a little worried about having all these dataset properties myself. It very well could lead misconfigurations by users. The read/write performance in general for non-recordsize aligned Direct IO is not very good, so I figured most people may just set:
- directio=on
- directio_read/write_align=record

I know @behlendorf and I were wanting to support non-recordsize aligned Direct IO as most user are accustomed to the requirement of PAGE_SIZE aligned requests for Direct IO as opposed to reccordsize alignment. Unfortunately, the non-recordsize aligned Direct IO requests currently incur a huge performance hit.

Reply:
One concern which came up was that some applications may use `read(2)` with `O_DIRECT` to probe for the alignment restriction in the underlying filesystem so they can do the right thing.  I'm told `qemu` may do this, but I'm not familiar with exactly how sophisticated it is about it.  There are probably other applications which do the same.

I think it's worth investigating at least the `qemu` behavior to make sure we avoid selecting a default behavior which can result in the alignment being incorrectly determined.

But I definitely agree we should see if we can distill down the possible number of settings to only what's actually useful.

Reply:
I do however think there is an argument to be made that if a user uses the O_DIRECT flag the expectation is that no caching will take place even if it comes with a performance penalty.

I know part of the reason we wanted to just stick with the PAGE_SIZE alignment requirement was to make sure that if a user doesn't want a memory copy to take place it will not happen. That is part of the reason for having so many dataset properties. I don't know if a user would exactly be too happy either if they requested O_DIRECT and we just ignored that request by default. There maybe specific reasons they do not want to cache any data.

It does get messy with so many properties, and we definitely could distill them a bit. I just think we should always respect the user's request to avoid caching if that's what they ask for. If they want to do no caching only when it is beneficial for performance that should also be an option as well.

Reply:
Is there a reason users want to avoid a memory copy other than performance?

Reply:
A main driver for using direct IO will be performance; however, there maybe users who just don’t want their data cached. 

DB’s that are doing their own caching schemes may just want to avoid a second copy in memory.

I imagine the majority of users will not using direct IO, and if they are, they will be cognizant of the record alignment. No matter what, people would not be happy if they asked to bypass caching and it still winds up in the ARC without them having any say (even if it was more performant to do so). We are trying to follow the fundamental concept of direct IO by bypass caching and deliver the data directly to the underlying device(s). I think leaving that decision in the users control is preferable over just deciding for them. That was the idea behind all the dataset properties. If the user does decide to write, read the data to, from the ARC when it is more beneficial to performance, even if direct IO was requested, then the properties will allow that to happen.

Reply:
> A main driver for using direct IO will be performance; however, there maybe users who just don’t want their data cached.

Is this use case not covered by the `primarycache` property?

I think the tricky case that we are discussing here is sequential reads (or writes) that are less than the recordsize.  e.g. recordisze=128K but the application is doing sequential reads in 4k chunks. I think there are logical arguments to be made either way. But the concern is that in this case if we do the reads directly (bypassing the ARC), we will do ~32x as much i/o as if we cached it.  So in this case directio is the opposite of "go fast mode".

Of course, applications should not do this, but we have to decide what ZFS will do if they try.  The options are:
1. do the i/o directly, taking a ~32x performance hit
2. do the i/o the normal way (ignore DIRECTIO request)
3. fail the i/o.

Reply:
Just form a user perspective, the many options propose seem like misconfiguration will be very likely so distilling them down sounds like a good idea...

But that aside, most of the values for the original proposal sort of make sense except maybe **legacy**, using **arc** here instead seems better.

As the directio requests will got through the arc per old behavior.

Reply:
Just to say that I agree with @ahrens. After all, even Linux own O_DIRECT support can fail over to pagecache in a number of cases, the most notable `mmap()` of an otherwise O_DIRECT opened file. You can see [`man open`](http://man7.org/linux/man-pages/man2/open.2.html) for other details. Moreover, if a user really want to bypass ARC in any case he can set `primarycache=none`, or I am missing something?

Reply:
> One concern which came up was that some applications may use read(2) with O_DIRECT to probe for the alignment restriction in the underlying filesystem so they can do the right thing. I'm told qemu may do this, but I'm not familiar with exactly how sophisticated it is about it. There are probably other applications which do the same.

I can't speak to this personally, but I did some research. Here is the code in question (and see the function after it, raw_probe_alignment() for more details):
https://github.com/qemu/qemu/blob/master/block/file-posix.c#L327

Specifically, if it ends up down this code path, it needs "EINVAL for misaligned O_DIRECT reads."